proposals/0000-decl-invis-binders.rst

```diff
@@ -0,0 +1,440 @@
+To cite the User's Guide, "The reason for this exception [the strange, ad-hoc
+rule about top-level kind annotations] is that there may be no other way to
+bind k".
+
+The same rule applies to type family instances::
+
+   type family F a :: k
+
+   type instance F Int = Any :: j -> j
+
+(where ``Any :: forall p. p``). This type instance is currently legal, but
+under this proposal ``j`` would not be in scope. You would have to write::
+
+   type instance F @(j->j) Int = Any :: j -> j
+
+In type instances, this form is already allowed today.
+
+Argument 5: Arity Inference
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+*Arity* is a property of type synonyms and type families that determines how
+many arguments are required at use sites (partial application is not allowed).
+The notion of arity is described in more detail in `section 6.4.9.2.1. "Type family
+declarations" <https://downloads.haskell.org/~ghc/9.2.1-alpha2/docs/html/users_guide/exts/type_families.html#type-family-declarations>`_
+of the User's Guide.
+
+Importantly, arity cannot be determined by looking at the kind of a type
+constructor. Consider ``F`` declared as follows::
+
+  type F :: Type -> forall k. Maybe k
+  type family F x
+
+The compiler can either assign it the arity of 1 or 2, and this choice will
+determine whether the equations of the type family can pattern match on ``k``.
+This will also determine whether a higher-kinded usage of ``F`` is possible.
+
+Thus, both arities can make sense depending on intended usage. Currently, GHC
+expects the programmer to employ an unsightly technique to specify the arity.
+By default, arity inference tries to include as many forall-bound variables as
+possible, to maximize the expressivity at definition site (at the cost of
+higher-kinded usage). However, the user may opt out by duplicating the return
+kind of the type family in its header::
+
+  -- arity = 2
+  type F :: Type -> forall k. Maybe k
+  type family F x
+
+  -- arity = 1
+  type G :: Type -> forall k. Maybe k
+  type family G x :: forall k. Maybe k
+
+With ``@``-binders we can do the opposite. We propose that by default, arity
+inference would include as few forall-bound variables as possible, to allow
+higher-kinded usage. However, it shall also include all ``@``-bound variables::
+
+  -- arity = 1
+  type F :: Type -> forall k. Maybe k
+  type family F x
+
+  -- arity = 2
+  type G :: Type -> forall k. Maybe k
+  type family G x @k
+
+This would simultaneously reduce code duplication and simplify the rules for
+arity inference.
+
+Motivation Summary
+~~~~~~~~~~~~~~~~~~
+
+Support for ``@k``-binders in type declarations will have the following
+positive effects:
+
+* Term- and type-level syntax become more consistent with one another
+* ``class`` and ``instance`` declarations can be more symmetric
+* Standalone kind signatures will no longer lead to duplication
+* Implicit quantification rules in type synonyms can be simplified
+* Arity inference becomes much more straightforward
+
```

Comment:
I think this proposal also strictly increases expressivity:

```hs
type family F a
type C :: forall a. F a -> Constraint
class C x where   -- no way to put a kind signature on x
  -- no way to bring a into scope here
```

Check me on the above, but I think there may be programs unwritable without this extension.

Relatedly, more simply, but more theoretically: This proposal means that every variable can have an explicit binding site. I'm pretty sure this is the last frontier (though I could be wrong).

Reply:
> This proposal means that every variable can have an explicit binding site. 

Yes!  Say this loud and clear!